### PR TITLE
Styled table group headers redux

### PR DESCRIPTION
**Pull request by savanah**
This patch allows custom styling of the headers for UITableViewStyleGrouped tables in the same way plain table group heads are styled with TTSTYLEVAR(tableHeaderTintColor)

Also adds to TTCatalog to demonstrate how it's used.
After a rebase, replaces pull request 439:  https://github.com/facebook/three20/pull/439

### CODE REVIEW

**Comment by jverkoey** [comment]
Ah damn, this is not going to make your day. These changes are going off of `facebook:master` instead of `facebook:development`.

To save you headache in the future, check out the commits that are included with your pull request before you send it off. I'm going to change the integration branch to `development`, so that should make this problem a lot less likely (I did the exact same thing last night by accident).

As for these pull requests, I should be able to pull your changes into development without including all of these commits, so don't worry about remaking the pull requests if it's a pain in the ass. Thanks for putting these together :)

**Comment by savanah** [comment]
I think I see what's going on.  I based my changes on facebook/development, but then when sending the pull request failed to change the "base branch" in the pull request.

Too bad github doesn't allow you (the upstream "facebook" repo) to specify which branch you'd like pull requests to be based on.

**Comment by savanah** [comment]
Heh, I typed "based my changes on facebook/development a38a6c4379315420fc5bd8efa52399a45080662d", but put square brackets around the SHA.  And _poof_ the square brackets and the SHA disappeared.

**Comment by jverkoey** [comment]
> Too bad github doesn't allow you (the upstream "facebook" repo) to specify which branch you'd like pull requests to be based on.

I'm going to make `development` the integration branch from now on. It might take a couple days because I've got to get Dave (our awesome Open Source guy at Facebook) to update the repo's settings.

File: src/Three20UI/Sources/TTTableViewDelegate.m [hunk 1]
```diff
@@ -86,29 +89,65 @@
...
+
+        } else if (tableView.style == UITableViewStyleGrouped && TTSTYLEVAR(tableHeaderGrouped)) {
+          TTLabel * label = [[[TTLabel alloc] initWithText:title] autorelease];
+          label.style = TTSTYLE(tableHeaderGrouped);
+          label.backgroundColor = [UIColor clearColor];
+          header = (TTTableHeaderView*)label;
+        }
+
+        if (header) {
           [_headers setObject:header forKey:title];
```
Comment (SpicaThor) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: Storing headers views to "cache" based on their titles making it very difficult to use when you have more than one section header with the same title on it. Such headers are drawing themselves only at last position, which kind of breaking the whole view... :(

Comment (savanah) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: Is there a reason not to use the section number as the key?  I guess it might be possible a single delegate would be responsible for more than one tableview...  maybe make a key out of the UITableView\* and the section number?

Still, this is bug compatible with current three20, so that problem is orthogonal to the feature addressed by this pull request.

Comment (SpicaThor) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: I think that the best way will be to implement the same caching mechanism as for UITableViewCell...

Comment (savanah) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: Can you expand on that?  

As I understand it, the UITableViewCell mechanism expects the delegate to group reusable cells by an identifier, and the table itself manages a queue of available cells that the delegate can request at the appropriate time.

Comment (SpicaThor) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: I don't think...
As I know the UITableView push reusable cell to it's inner reusable cells cache based on it's identifier. Then the cell is released. It's not vanished because of the cache that "holds" it till some UITableViewDelegate requests for one.(or till the cache itself is released by UITableView) Now, for headers the UITableView doesn't do anything like that. If you mean to use some subclass of UITableViewCell as a header view, it's cool, but it's probably not going to help us as UITableView doesn't push the headers views into cache and so it's impossible to get it from there after a while. As for the regular table view cells the UITableView moves them to the cache after they become invisible, but as long, as we don't know their algorithm we can not know, that it will do that for cells we put in other places.(headers, footers and so)

Comment (savanah) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: Kind of hard to parse your response, but I think we're in agreement that the UITableViewCell caching mechanism is of no help here.

... and the key thing we need to come back to is that this issue of caching header views by title is unrelated to this pull request.

Coming up with a new way of caching headers (perhaps a setting on the delegate or datasource skips caching altogether?) sounds like the work of another branch / patch / pull request.

Comment (SpicaThor) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: That's true.
Btw, I'v already made a new request, so generally we can move this conversation...

P.S. The delegate idea sounds nice to me.

Comment (SpicaThor) [inline src/Three20UI/Sources/TTTableViewDelegate.m:147]: https://github.com/facebook/three20/issues/544

**Comment by rikkert** [comment]
Why are you using a TTStyle for the tableHeaderGrouped var and returning a TTLabel?
Isn't it more persistent to use TTTableHeaderView with perhaps different style vars for a grouped header?
@jverkoey when can we expect this to be merged into master?

**Comment by jwang** [comment]
@savanah - were you able to address the questions and issues? I want to pull this in for testing next week.

**Comment by rikkert** [comment]
@jwang any eta on pulling this one in?

**Comment by jamesgpearce** [comment]
Facebook has not maintained or supported Three20 for some time, and we are closing its old and outstanding pull requests.

Many, many thanks for your support of the project. If you have any further questions, please don't hesitate to let me know.
